### PR TITLE
feat(contracts-periphery): Add Optimist SBT deploy script

### PR DESCRIPTION
**Pull request by roninjin10**
- Add deploy script for optimist sbt

### CODE REVIEW

**Comment by roninjin10** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4372** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/4372" target="_blank">Graphite</a> 
    * **PR #4370** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/4370" target="_blank">Graphite</a> 
      * **PR #4371** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/4371" target="_blank">Graphite</a> 
        * **PR #4373** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/4373" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/4373?utm_source=stack-comment).

**Comment by roninjin10** [comment]
this is more or less ready to go I think

**Comment by roninjin10** [comment]
nvm depends on attestation deployment merging first

File: packages/contracts-periphery/deploy/op-nft/Optimist.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  ].deployments.get('AttestationStation')
+
+  const attestationStationAddress = Deployment__AttestationStation.address
+
+  const { deploy } = await hre.deployments.deterministic('Optimist', {
+    salt: hre.ethers.utils.solidityKeccak256(['string'], ['Optimist']),
+    from: deployer,
+    // make these more configurable
+    args: [
+      'OptimistNFT',
```
Comment (tynes) [inline packages/contracts-periphery/deploy/op-nft/Optimist.ts:18]: The hardhat deploy config package can help with making these args configurable

File: packages/contracts-periphery/deploy/op-nft/OptimistImpl.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  const Deployment__AttestationStation = await hre.deployments.get(
+    'AttestationStationProxy'
+  )
+  const attestationStationAddress = Deployment__AttestationStation.address
+
+  const { deploy } = await hre.deployments.deterministic('Optimist', {
+    salt: hre.ethers.utils.solidityKeccak256(['string'], ['Optimist']),
+    from: deployer,
+    // make these more configurable
```
Comment (tynes) [inline packages/contracts-periphery/deploy/op-nft/OptimistImpl.ts:15]: Can use the `hardhat-deploy-config` plugin, we use it elsewhere for deployments

File: packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    contract: 'Proxy',
+    args: [proxyAdmin],
+    postDeployAction: async (contract) => {
+      await assertContractVariable(contract, 'admin', proxyAdmin)
+    },
+  })
+
+  const Deployment__OptimistProxy = await hre.deployments.get('OptimistProxy')
+  console.log(`OptimistProxy deployed to ${Deployment__OptimistProxy.address}`)
+}
```
Comment (tynes) [inline packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts:27]: I think you should use the deployer as the initial admin and then set the impl and then change the admin to the proxy admin

File: packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    const receipt = await tx.wait()
+    console.log(`implementation set in ${receipt.transactionHash}`)
+  } else {
+    console.log('implementation already set to Optimist contract')
+  }
+
+  const l2ProxyOwnerAddress = hre.deployConfig.l2ProxyOwnerAddress
+  const admin = await Proxy.callStatic.admin()
+  console.log(`admin set to ${admin}`)
+  if (getAddress(admin) === getAddress(l2ProxyOwnerAddress)) {
```
Comment (tynes) [inline packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts:72]: This needs to be `!==` not `===`

File: packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    console.log('admin not set correctly')
+    console.log(`Setting admin to ${l2ProxyOwnerAddress}`)
+
+    const tx = await Proxy.changeAdmin(l2ProxyOwnerAddress)
+    const receipt = await tx.wait()
+    console.log(`admin set in ${receipt.transactionHash}`)
+  } else {
+    console.log('admin already set to L2 multisig')
```
Comment (tynes) [inline packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts:80]: this log needs to be updated

Comment (roninjin10) [inline packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts:80]:
```suggestion
    console.log('admin already set to Optimist')
```

File: packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts (hunk 4 above)
Comment (roninjin10) [inline packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts:72]:
```suggestion
  if (getAddress(admin) !== getAddress(l2ProxyOwnerAddress)) {
```

**Review by roninjin10 — COMMENTED** [review]
lgtm

File: packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  console.log(`admin set to ${admin}`)
+  if (getAddress(admin) !== getAddress(l2ProxyOwnerAddress)) {
+    console.log('admin not set correctly')
+    console.log(`Setting admin to ${l2ProxyOwnerAddress}`)
+
+    const tx = await Proxy.changeAdmin(l2ProxyOwnerAddress)
+    const receipt = await tx.wait()
+    console.log(`admin set in ${receipt.transactionHash}`)
+  } else {
+    console.log('admin already set to Optimist')
```
Comment (tynes) [inline packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts:80]: This comment seems off, should be `admin already set to proxy owner address`

File: packages/contracts-periphery/src/config/deploy.ts [hunk 7]
```diff
@@ -14,14 +14,31 @@ export interface DeployConfig {
...
+   * Number of confs before considering it final
    */
-  l2ProxyOwnerAddress: string
+  numDeployConfirmations?: number
 
   /**
-   * Number of confs before considering it final
+   * Name of the NFT in the Optimist contract.
    */
-  numDeployConfirmations?: number
```
Comment (tynes) [inline packages/contracts-periphery/src/config/deploy.ts:24]: We still need `numDeployConfirmations`

File: packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts (hunk 6 above)
Comment (roninjin10) [inline packages/contracts-periphery/deploy/op-nft/OptimistProxy.ts:80]:
```suggestion
    console.log('admin already set to proxy owner address')
```

**Pull request merged**